File: filly-core/src/widgets/gauge.rs
use crate::event::Event;
use crate::render::{
    Alignment, BorderStyle, EdgeInsets, Rect, RenderNode, RenderTree, TextStyle,
};
use crate::widget::{EventResult, Widget};
use filly_protocol::WidgetResponse;

pub struct GaugeWidget {
    pub title: String,
    pub percent: u16,
    pub label: String,
    dirty: bool,
}

impl GaugeWidget {
    pub fn new(title: String, percent: u16, label: String) -> Self {
        Self {
            title,
            percent,
            label,
            dirty: true,
        }
    }
}

impl Widget for GaugeWidget {
    fn render(&self, area: Rect) -> RenderTree {
        let box_w = 50u16.min(area.width.saturating_sub(2));
        let box_h = 6u16.min(area.height.saturating_sub(2));
        let box_x = (area.width.saturating_sub(box_w)) / 2;
        let box_y = (area.height.saturating_sub(box_h)) / 2;

        let mut children = Vec::new();
        children.push(RenderNode::Text {
            rect: Rect::new(1, 0, box_w - 2, 1),
            content: self.title.clone(),
            alignment: Alignment::Center,
            style: TextStyle {
                bold: true,
                fg: Some("green".into()),
                ..TextStyle::normal()
            },
            role: None,
            label: None,
            description: None,
        });
        children.push(RenderNode::Gauge {
            rect: Rect::new(1, 2, box_w - 2, 2),
            percent: self.percent,
            label: self.label.clone(),
            style: TextStyle::accent(),
            role: None,
            description: None,
        });
        children.push(RenderNode::Text {
            rect: Rect::new(1, box_h - 1, box_w - 2, 1),
            content: "Any key to dismiss".into(),
            alignment: Alignment::Center,
            style: TextStyle::muted(),
            role: None,
            label: None,
            description: None,
        });

        RenderTree::Container {
            rect: Rect::new(box_x, box_y, box_w, box_h),
            background: None,
            border: BorderStyle::Single,
            padding: EdgeInsets::zero(),
            children,
            role: None,
            label: None,
            description: None,
        }
    }

    fn handle_event(&mut self, event: &Event) -> EventResult {
        if let Event::Key(_, _) = event {
            EventResult::Response(WidgetResponse {
                result: None,
                cancelled: false,
                error: None,
            })
        } else {
            EventResult::Unhandled
        }
    }

    fn is_dirty(&self) -> bool {
        self.dirty
    }
    fn clear_dirty(&mut self) {
        self.dirty = false;
    }
}

File: filly-core/widgets/gauge.c
#include "gauge.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

typedef struct {
    char *title;
    int percent;
    char *label;
    bool dirty;
} GaugeData;

static void gauge_render(Widget *self, Rect area, RenderTree *out) {
    GaugeData *d = (GaugeData *)(self + 1);
    memset(out, 0, sizeof(*out));
    int box_w = 50, box_h = 6;
    if (box_w > area.w - 2) box_w = area.w - 2;
    int box_x = (area.w - box_w) / 2, box_y = (area.h - box_h) / 2;

    RenderTree *children = calloc(3 + 1, sizeof(RenderTree));
    children[0].type = RNODE_TEXT;
    children[0].rect = rect_new(1, 0, box_w - 2, 1);
    children[0].text.content = strdup(d->title);
    children[0].text.align = ALIGN_CENTER;
    children[0].text.style = textstyle_selected();

    children[1].type = RNODE_GAUGE;
    children[1].rect = rect_new(1, 2, box_w - 2, 2);
    children[1].gauge.percent = d->percent;
    children[1].gauge.label = strdup(d->label);
    children[1].gauge.style = textstyle_accent();

    children[2].type = RNODE_TEXT;
    children[2].rect = rect_new(1, box_h - 2, box_w - 2, 1);
    children[2].text.content = strdup("Any key to dismiss");
    children[2].text.align = ALIGN_CENTER;
    children[2].text.style = textstyle_muted();

    out->type = RNODE_CONTAINER;
    out->rect = rect_new(box_x, box_y, box_w, box_h);
    out->container.border = BORDER_SINGLE;
    out->container.padding = edgeinsets_zero();
    out->container.children = children;
    out->container.child_count = 3;
}

static EventResult gauge_handle_event(Widget *self, Event *ev, Backend *backend) {
    (void)self;
    (void)backend;
    if (ev->type == EVENT_KEY)
        return event_result_response((WidgetResponse){ .result = NULL, .cancelled = false, .error = NULL });
    return event_result_unhandled();
}

static bool gauge_is_dirty(Widget *self) { return ((GaugeData *)(self + 1))->dirty; }
static void gauge_clear_dirty(Widget *self) { ((GaugeData *)(self + 1))->dirty = false; }
static void gauge_destroy(Widget *self) {
    GaugeData *d = (GaugeData *)(self + 1);
    free(d->title); free(d->label);
}

Widget *gauge_widget_new(const char *title, int percent, const char *label) {
    Widget *w = calloc(1, sizeof(Widget) + sizeof(GaugeData));
    w->vtable.render = gauge_render;
    w->vtable.handle_event = gauge_handle_event;
    w->vtable.is_dirty = gauge_is_dirty;
    w->vtable.clear_dirty = gauge_clear_dirty;
    w->vtable.destroy = gauge_destroy;
    GaugeData *d = (GaugeData *)(w + 1);
    d->title = strdup(title);
    d->percent = percent;
    d->label = strdup(label);
    d->dirty = true;
    return w;
}